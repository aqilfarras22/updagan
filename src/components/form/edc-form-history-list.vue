<template>
  <table-component
    showEmpty
    noGutters
    :items="history"
    :fields="formHistoryFields"
  >
    <!-- date -->
    <template slot="createdAt" slot-scope="data">
      {{ new Date(data.row.item.createdAt) | formatUTCDate }}
    </template>
    <!-- action -->
    <template slot="eventType" slot-scope="data">
      {{ data.row.item.primaryReason && data.row.item.primaryReason!='' ? data.row.item.primaryReason : (verificationStatusText[data.row.item.eventType] || approvalStatusText[data.row.item.eventType] || historyStatusText[data.row.item.eventType]) }}
    </template>
    <!-- view -->
    <template slot="view" slot-scope="data">
      <a v-if="historyStatus.MODIFIED === data.row.item.eventType" @click.prevent="view(data.row.item)" href="">View</a>
    </template>
  </table-component>
</template>

<script>
import TableComponent from '@/components/common/table';
import {
  approvalStatusText,
  verificationStatusText,
  historyStatusText,
  historyStatus,
} from '@/constants/form';

export default {
  name: 'form-history-list',
  components: {
    TableComponent,
  },
  props: {
    history: {
      type: Array,
      default: null,
    },
  },
  data() {
    return {
      approvalStatusText,
      verificationStatusText,
      historyStatusText,
      historyStatus,
      formHistoryFields: [
        {
          key: 'createdAt',
          label: 'Date',
          sortable: true,
        },
        {
          key: 'eventType',
          label: 'Action / Changes',
          sortable: true,
        },
        {
          key: 'createdBy',
          label: 'By',
          sortable: true,
        },
        {
          key: 'view',
          label: ' ',
        },
      ],
    };
  },
  methods: {
    view(item) {
      this.$emit('view', item);
    },
  },
};
</script>
