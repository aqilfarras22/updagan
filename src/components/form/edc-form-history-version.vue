<template>
  <div>
    <b-row class="border-bottom header-row">
      <b-col><a @click.prevent="back" class="backButton" href=""><i class="fa fa-caret-left fa-lg"></i>&nbsp; Back to history</a></b-col>
      <b-col class="text-right"><i class="square"></i>&nbsp; <span class="info">Edited fields</span></b-col>
    </b-row>
    <div>
      <b-card-group v-if="!edcFormVersionData.isUnSchedule" class="pb-15">
        <b-card no-body class="border-0" :header="edcFormVersionData.name">
          <div v-if="edcFormVersionData.primaryReason" class="reason">
            Reason for change: {{ edcFormVersionData.primaryReason }}
          </div>
          <SubmittedFormPreview v-if="edcFormVersionData.fields" :isHistory="true" :formData="edcFormVersionData.fields" />
        </b-card>
      </b-card-group>
    </div>
    <div v-if="edcFormVersionData.isUnSchedule && edcFormVersionData.fields">
      <b-card-group v-for="(unscheduleAnswers, index) in edcFormVersionData.fields" :key="index" :class="{'border-top': index!=0 ,'pb-15' : index+1 === edcFormVersionData.fields.length, 'pb-5' : index+1 !== edcFormVersionData.fields.length}">
        <b-card no-body class="border-0" :header="unscheduleAnswers.formName">
          <div v-if="edcFormVersionData.primaryReason" class="reason">
            Reason for change: {{ edcFormVersionData.primaryReason }}
          </div>
          <SubmittedFormPreview v-if="unscheduleAnswers.formData" :isHistory="true" :formData="unscheduleAnswers.formData"/>
        </b-card>
      </b-card-group>
    </div>
  </div>
</template>

<script>
import SubmittedFormPreview from '@/components/edc-form/submitted-form-preview';

export default {
  name: 'form-history-version',
  components: { SubmittedFormPreview },
  props: {
    edcFormVersionData: Object,
  },
  methods: {
    back() {
      this.$emit('back');
    },
  },
};
</script>

<style scoped>
.header-row {
  padding: 0.9rem 0 0.65rem;
}
.square {
  display: inline-block;
  width: 16px;
  height: 16px;
  border-radius: 1px;
  margin-bottom: -2px;
  background-color: #ffa200;
}
.info {
  font-size: 12px;
  display: inline-block;
}
.reason {
  font-family: "RobotoBold" !important;
  color: #666669;
  padding-left: 30px;
  padding-top: 19px;
}
</style>
