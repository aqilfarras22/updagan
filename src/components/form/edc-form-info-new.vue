<template>
  <div class="edc-form-bar">
	<div class="row">
		<div class="col-md-2">
			<div class="heading">Form/Packet Name</div>
			<div class="value">{{ clearFormName() }}</div>
		</div>
		<div class="col-md-2">
			<div class="heading">Verification Status</div>
			<div class="value" :class="edcFormInfo.verificationStatus">{{ verificationStatusText[edcFormInfo.verificationStatus] }}</div>
			<div v-if="verified && edcFormInfo.verifiedOn" class="date value">
				({{ new Date(edcFormInfo.verifiedOn) | formatGMTDate }})
			</div>
		</div>
		<div class="col-md-2">
			<div class="heading">Approval Status</div>
			<div class="value" :class="edcFormInfo.approvalStatus">{{ approvalStatusText[edcFormInfo.approvalStatus] }}</div>
			<span v-if="approved && edcFormInfo.approvedOn" class="date value">
			({{ new Date(edcFormInfo.approvedOn) | formatGMTDate }})
			</span>
		</div>
	</div>
  </div>
</template>

<script>
import { AclRule } from 'vue-acl';
import { mapGetters } from 'vuex';
import {
  approvalStatusText,
  verificationStatusText,
  verificationStatus,
  approvalStatus,
} from '@/constants/form';

export default {
  name: 'form-info',
  components: {},
  props: {
    edcFormInfo: {
		type: Object,
		default: {}
	}
  },
  data() {
    return {
      approvalStatusText,
	  verificationStatusText,
	  verificationStatus
	};
  },
  computed: {
    ...mapGetters(['user']),
    userAccess() {
      return new AclRule(this.user.userRole).generate();
    }
  },
  methods: {
	clearFormName() {
    	return this.edcFormInfo.formName ? this.edcFormInfo.formName.replace(/_/g, ' ') : '';
	},
	verified() {
      	return this.edcFormInfo.verificationStatus === verificationStatus.VERIFIED;
    },
    approved() {
      	return this.edcFormInfo.approvalStatus === approvalStatus.APPROVED;
    }
  }
};
</script>

<style scoped lang="scss">
.edc-form-bar {
	height: 60px;
  background-color: #fff;
  padding: 10px 0px;
  -webkit-box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.06);
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.06);
  margin-top: 15px;
  height: 75px;
  .heading {
    font-weight: 600;
  }
}
.date {
  font-size: 12px;
  line-height: 1.67;
}
.telehealth-call {
  .heading {
    font-size: 10px;
  }
  .value {
    font-size: 10px;
  }
}
</style>
