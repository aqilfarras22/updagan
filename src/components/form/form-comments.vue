<template>
  <div>
    <b-card-group class="border-top">
      <b-card no-body class="border-0" :header="title">
        <b-list-group v-if="comments" flush>
          <b-list-group-item class="item border-0" v-for="comment in comments" :key="comment.id">
            <b-row>
              <div class="col-5">
                <div class="name-class pb-5">
                  <span v-if="comment.author.name.firstName">{{ comment.author.name.firstName }}</span>
                  <span v-if="comment.author.name.middleName">{{ comment.author.name.middleName }}</span>
                  <span v-if="comment.author.name.lastName">{{ comment.author.name.lastName }}</span>
                </div>
                <div>{{ new Date(comment.postedAt * 1000 ) | formatGMTDate }}</div>
              </div>
              <div class="col-7 div_style">{{ comment.message }}</div>
            </b-row>
          </b-list-group-item>
        </b-list-group>
      </b-card>
    </b-card-group>
  </div>
</template>

<script>
import ButtonCmp from '@/components/common/button';

export default {
  name: 'form-fields-list',
  components: { ButtonCmp },
  props: {
    comments: Array,
  },
  data() {
    return {
      title: 'Comments',
    };
  },
  methods: {},
};
</script>

<style scoped>
.item {
  font-size: 0.9rem;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.43;
  text-align: left;
  color: #666669;
  padding-left: 30px;
}
.name-class {
  font-weight: 700;
}
.div_style {
  border-left: 1px solid #d4d9e2; 
}
</style>
