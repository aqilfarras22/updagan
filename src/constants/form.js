export const fieldTypes = {
  POLAR_QUESTION: 'POLAR_QUESTION',
  DROPDOWN: 'DROPDOWN',
  DATEPICKER: 'DATEPICKER',
  INPUT: 'INPUT',
  LABEL: 'LABEL',
};

export const formFieldQueryStatus = {
  OPEN: 'OPEN',
  CLOSE: 'CLOSE',
};

export const historyStatus = {
  MODIFIED: 'MODIFIED',
};

export const historyStatusText = {
  MODIFIED: 'Original Submission',
};

// approve
export const approvalStatus = {
  APPROVED: 'APPROVED',
  NOT_APPROVED: 'NOT_APPROVED',
};
export const approvalStatusText = {
  APPROVED: 'Approved',
  NOT_APPROVED: 'Unapproved',
};

// verification
export const verificationStatus = {
  VERIFIED: 'VERIFIED',
  NOT_VERIFIED: 'NOT_VERIFIED',
};
export const verificationStatusText = {
  VERIFIED: 'Verified',
  NOT_VERIFIED: 'Unverified',
};
