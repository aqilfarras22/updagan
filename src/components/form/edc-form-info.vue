<template>
  <div v-if="formId">
    <dl v-if="formName" ref="formName">
      <dt>Form/Packet Name</dt>
      <dd>{{ clearFormName }}</dd>
    </dl>
    <dl ref="verificationStatus">
      <dt>Verification Status</dt>
      <dd>
        <div :class="verificationStatus">{{ verificationStatusText[verificationStatus] }}</div>
        <div v-if="verified && verifiedOn" class="date">
          ({{ new Date(verifiedOn) | formatGMTDate }})
        </div>
      </dd>
    </dl>
    <dl ref="approvalStatus">
      <dt>Approval Status</dt>
      <dd>
        <div :class="approvalStatus">{{ approvalStatusText[approvalStatus] }}</div>
        <span v-if="approved && approvedOn" class="date">
          ({{ new Date(approvedOn) | formatGMTDate }})
        </span>
      </dd>
    </dl>
    <div class="actions">
      <button-cmp v-if="canPiAccess || canSubPiAccess || canScAccess || canHomeHealthAccess" class="form-action-button" size="lg" variant="primary" @click="openApprovalModal">{{ approvalBtnText }}</button-cmp>
      <button-cmp v-if="canCraAccess" class="form-action-button" size="lg" variant="primary" @click="openVerificationModal">{{ verifyBtnText }}</button-cmp>
    </div>
    <formApprovalModal
      v-if="canPiAccess || canSubPiAccess || canScAccess || canHomeHealthAccess"
      ref="approvalModal"
      callApi
      :formId="formId"
      :approved="approved"
      @confirmed="approvalConfirmed"
    />
    <formVerificationModal
      v-if="canCraAccess"
      ref="verificationModal"
      callApi
      :formId="formId"
      :verified="verified"
      @confirmed="verificationConfirmed"
    />
  </div>
</template>

<script>
import { AclRule } from 'vue-acl';
import { mapGetters } from 'vuex';
import formVerificationModal from '@/components/modals/edc-form-verification-modal';
import formApprovalModal from '@/components/modals/edc-form-approval-modal';
import ButtonCmp from '@/components/common/button';

import { PI, CRA, SUB_PI, STUDY_COORDINATOR, HOME_HEALTH } from '@/constants/authRoles';
import {
  approvalStatusText,
  verificationStatusText,
  verificationStatus,
  approvalStatus,
} from '@/constants/form';

export default {
  name: 'form-info',
  components: {
    formVerificationModal,
    formApprovalModal,
    ButtonCmp,
  },
  props: {
    formId: String,
    formName: String,
    approvalStatus: String,
    verificationStatus: String,
    verifiedOn: [String, Number],
    approvedOn: [String, Number],
  },
  data() {
    return {
      approvalStatusText,
      verificationStatusText,
    };
  },
  computed: {
    ...mapGetters(['user']),
    userAccess() {
      return new AclRule(this.user.userRole).generate();
    },
    canPiAccess() {
      this.$acl.change(`${PI}`);
      return this.$acl.check('userAccess');
    },
    canSubPiAccess() {
      this.$acl.change(`${SUB_PI}`);
      return this.$acl.check('userAccess');
    },
    canScAccess() {
      this.$acl.change(`${STUDY_COORDINATOR}`);
      return this.$acl.check('userAccess');
    },
    canHomeHealthAccess() {
      this.$acl.change(`${HOME_HEALTH}`);
      return this.$acl.check('userAccess');
    },
    canCraAccess() {
      this.$acl.change(`${CRA}`);
      return this.$acl.check('userAccess');
    },
    clearFormName() {
      return this.formName ? this.formName.replace(/_/g, ' ') : '';
    },
    verified() {
      return this.verificationStatus === verificationStatus.VERIFIED;
    },
    approved() {
      return this.approvalStatus === approvalStatus.APPROVED;
    },
    verifyBtnText() {
      return this.verified ? 'Unverify' : 'Verify';
    },
    approvalBtnText() {
      return this.approved ? 'Unapprove' : 'Approve';
    },
  },
  methods: {
    openVerificationModal() {
      this.$refs.verificationModal.show();
    },
    openApprovalModal() {
      this.$refs.approvalModal.show();
    },
    verificationConfirmed({ verified }) {
      const status = verified ? verificationStatus.VERIFIED : verificationStatus.NOT_VERIFIED;
      this.$emit('verificationConfirmed', status);
    },
    approvalConfirmed({ approved }) {
      const status = approved ? approvalStatus.APPROVED : approvalStatus.NOT_APPROVED;
      this.$emit('approvalConfirmed', status);
    },
  },
};
</script>

<style scoped>
.APPROVED,
.VERIFIED {
  color: #48bf7a !important;
}
.NOT_VERIFIED,
.NOT_APPROVED {
  color: #f16559 !important;
}
.form-action-button {
  width: 100%;
  margin-top: 1px;
  margin-bottom: 1px;
}
dt {
  font-size: 10px;
  line-height: 1.6;
  text-transform: uppercase;
  color: #adaeb1;
}
dd {
  color: #666669;
  font-size: 14px;
  font-weight: 500;
}
.date {
  font-size: 12px;
  line-height: 1.67;
}
</style>
