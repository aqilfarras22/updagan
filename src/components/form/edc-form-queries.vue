<template>
  <b-tabs class="" v-model="tabIndex">
    <b-tab :title="`Open Queries: ${openQueries.length}`" :title-link-class="queriesOpen()">
      <FormQueriesCommentForm v-if="value" @cancel="cancelQuery" @submit="createQuery"/>
      <div v-else>
        <div v-if="openQueryId">
          <FormQueriesQueryDetails
            :userNames="userNames"
            :comments="comments"
            :query="openQuery"
            @back="hideOpenQuery"
            @onQueryCommentCreate="onQueryCommentCreate"
            @onQueryClosed="onQueryClosed"
          />
        </div>
        <div v-else>
          <FormQueriesHeader v-if="filter" :title="`Filtered by field ID ${shortFilterId}`" @back="resetFilter" backTitle="Clear"/>
          <b-row v-if="!openQueries.length" class="py-2">
            <b-col><p>No queries created yet.<br>Select a field to create a query.</p></b-col>
          </b-row>
          <FormQueriesList
            v-else
            :queries="openQueries"
            @view="viewOpenQuery"
          />
        </div>
      </div>
    </b-tab>
    <b-tab :title="`Closed Queries: ${closedQueries.length}`">
      <div v-if="closedQuery">
        <FormQueriesQueryDetails
          closed
          :comments="comments"
          :userNames="userNames"
          :query="closedQuery"
          @back="hideClosedQuery"
          @onQueryReopened="onQueryReopened"
        />
      </div>
      <div v-else>
        <FormQueriesHeader v-if="filter" :title="`Filtered by field ID ${filter}`" @back="resetFilter" backTitle="Clear"/>
        <b-row v-if="!closedQueries.length">
          <b-col class="py-2"><p>No queries closed yet.</p></b-col>
        </b-row>
        <FormQueriesList
          v-else
          :queries="closedQueries"
          @view="viewClosedQuery"
        />
      </div>
    </b-tab>
  </b-tabs>
</template>

<script>
import sortBy from 'lodash/sortBy';
import FormQueriesList from '@/components/form/edc-form-queries/form-queries-list';
import FormQueriesQueryDetails from '@/components/form/edc-form-queries/form-queries-query-details';
import FormQueriesCommentForm from '@/components/form/edc-form-queries/form-queries-comment-form';
import FormQueriesHeader from '@/components/form/form-queries-header';

import { formFieldQueryStatus } from '@/constants/form';

export default {
  name: 'edc-form-queries',
  components: {
    FormQueriesList,
    FormQueriesQueryDetails,
    FormQueriesCommentForm,
    FormQueriesHeader,
  },
  props: {
    value: String,
    queries: Array,
    filter: String,
    formId: String,
    studyId: String,
    comments: Array,
  },
  data() {
    return {
      tabIndex: 0,
      openQueryId: null,
      closedQueryId: null,
      userNames: {},
    };
  },
  async mounted() {
    const type="studyteam"
    const associated = "true"
    const studyTeamData = await this.$api.getStudyUsersFact(this.$route.params.studyId, type, associated);
    const siteTeamData = await this.$api.getUserNames(this.$route.params.studyId);
    studyTeamData.body.forEach(team => {
      this.userNames[team.id] = `${team.firstName} ${team.lastName}`;
    });
    siteTeamData.body.forEach(team => {
      this.userNames[team.id] = `${team.firstName} ${team.lastName}`;
    });
  },
  computed: {
    shortFilterId() {
      if (this.filter) {
        // return this.filter.slice(0, 5).toUpperCase();
        return this.filter;
      }
      return '';
    },
    filteredQueries() {
      let filtered = this.queries;
      if (this.filter) {
        filtered = this.queries.filter(q => q.fieldId === this.filter);
        this.$forceUpdate();
      }
      return sortBy(filtered, ['openedOn']).reverse();
    },
    openQueries() {
      return this.filteredQueries.filter(q => q.status === formFieldQueryStatus.OPEN);
    },
    closedQueries() {
      return this.filteredQueries.filter(q => q.status !== formFieldQueryStatus.OPEN);
    },
    openQuery() {
      return this.queries.find(q => q.queryId === this.openQueryId);
    },
    closedQuery() {
      return this.queries.find(q => q.queryId === this.closedQueryId);
    },
  },
  watch: {
    value(val) {
      if (val) {
        this.openQueryId = null;
        this.closedQueryId = null;
        this.tabIndex = 0;
      }
    },
    filter(val) {
      if (!this.value) {
        if (val) {
          this.shortFilterId = val;
          this.openQueryId = null;
          this.closedQueryId = null;
          this.tabIndex = 0;
        }
      }
    },
  },
  methods: {
    queriesOpen() {
      return this.filteredQueries.filter(q => q.status === formFieldQueryStatus.OPEN).length > 0 ? 'text-warning' : '';
    },
    viewOpenQuery(queryId) {
      this.openQueryId = queryId;
      const field = this.openQueries.find(q => q.queryId === queryId);
      this.$emit('updateComment', queryId, field.fieldId);
    },
    hideOpenQuery() {
      this.$emit('updateQueries');
      this.openQueryId = null;
      this.$emit('resetFilter');
    },
    viewClosedQuery(queryId) {
      this.closedQueryId = queryId;
      const field = this.closedQueries.find(q => q.queryId === queryId);
      this.$emit('updateComment', queryId, field.fieldId);
    },
    hideClosedQuery() {
      this.closedQueryId = null;
      this.$emit('resetFilter');
    },
    createQuery(commentData) {
      const newQuery = {
        fieldId: this.value,
        formId: this.formId,
        comment: commentData,
      };
      this.$emit('createQuery', newQuery);
      this.$emit('input', null);
    },
    onQueryCommentCreate(commentData) {
      this.$emit('createComment', commentData);
    },
    cancelQuery() {
      this.$emit('input', null);
      this.$emit('resetFilter');
    },
    resetFilter() {
      this.$emit('resetFilter');
      window.scrollTo(0, 0);
    },
    onQueryClosed(query) {
      this.openQueryId = null;
      this.$emit('onQueryClosed', query);
    },
    onQueryReopened(query) {
      this.closedQueryId = null;
      this.$emit('onQueryReopened', query);
    },
  },
};
</script>
<style>
a.text-warning:hover, a.text-warning:focus, .text-warning {
  color: #f5a623 !important;
}
</style>
