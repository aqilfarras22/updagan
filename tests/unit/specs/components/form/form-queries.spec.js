import BootstrapVue from 'bootstrap-vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';

import formQueriesComponent from '@/components/form/form-queries';
import { formFieldQueryStatus } from '@/constants/form';

const { it, describe, expect } = global;

const testPropsData = {
  value: 'model',
  queries: [
    {
      queryId: '1',
      fieldId: '11111',
      openedOn: 1,
      status: formFieldQueryStatus.OPEN,
    },
    {
      queryId: '2',
      fieldId: '11111',
      openedOn: 0,
      status: formFieldQueryStatus.OPEN,
    },
    {
      queryId: '3',
      fieldId: '33333',
      openedOn: 2,
      status: formFieldQueryStatus.CLOSE,
    },
  ],
  filter: '',
  formId: '1',
  fakeProp: '',
};

const localVue = createLocalVue();
localVue.use(BootstrapVue);

describe('components/form/form-queries.vue', () => {
  const createMountedWrapper = (options = {}) => shallowMount(formQueriesComponent, {
    localVue,
    propsData: { ...testPropsData },
    ...options,
    stubs: {
      FormQueriesList: {
        name: 'FormQueriesList',
        render: h => h('b', 'FormQueriesList'),
      },
      FormQueriesQueryDetails: {
        name: 'FormQueriesQueryDetails',
        render: h => h('b', 'FormQueriesQueryDetails'),
      },
      FormQueriesCommentForm: {
        name: 'FormQueriesCommentForm',
        render: h => h('b', 'FormQueriesCommentForm'),
      },
      FormQueriesHeader: {
        name: 'FormQueriesHeader',
        render: h => h('b', 'FormQueriesHeader'),
      },
    },
  });
  it('should create instance', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it('should have described child components', () => {
    expect(formQueriesComponent.components.FormQueriesList).toBeTruthy();
    expect(formQueriesComponent.components.FormQueriesQueryDetails).toBeTruthy();
    expect(formQueriesComponent.components.FormQueriesCommentForm).toBeTruthy();
    expect(formQueriesComponent.components.FormQueriesHeader).toBeTruthy();
  });

  it('should able to receive props', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.fakeProp).toBeUndefined();
    expect(wrapper.vm.value).toBe(testPropsData.value);
    expect(wrapper.vm.queries).toEqual(testPropsData.queries);
    expect(wrapper.vm.filter).toBe(testPropsData.filter);
    expect(wrapper.vm.formId).toBe(testPropsData.formId);
  });

  it('should have all required initial data', () => {
    const wrapper = createMountedWrapper();
    // initial data
    expect(wrapper.vm.tabIndex).toBe(0);
    expect(wrapper.vm.openQueryId).toEqual(null);
    expect(wrapper.vm.closedQueryId).toEqual(null);
  });

  it('should return computed property: shortFilterId', () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({ filter: 'qwertyasdfg' });
    // first 5 charts in uppercase
    expect(wrapper.vm.shortFilterId).toBe('QWERT');
    wrapper.setData({ filter: 'zxcvbasdfg' });
    expect(wrapper.vm.shortFilterId).toBe('ZXCVB');
  });

  it('should return computed property: filteredQueries', () => {
    const wrapper = createMountedWrapper();
    // from new to old
    expect(wrapper.vm.filteredQueries).toEqual([
      testPropsData.queries[2],
      testPropsData.queries[0],
      testPropsData.queries[1],
    ]);
    wrapper.setData({ filter: '33333' });
    expect(wrapper.vm.filteredQueries).toEqual([testPropsData.queries[2]]);
    wrapper.setData({ filter: '11111' });
    expect(wrapper.vm.filteredQueries).toEqual([
      testPropsData.queries[0],
      testPropsData.queries[1],
    ]);
  });

  it('should return computed property: openQueries', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.openQueries).toEqual([testPropsData.queries[0], testPropsData.queries[1]]);
  });

  it('should return computed property: closedQueries', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.closedQueries).toEqual([testPropsData.queries[2]]);
  });

  it('should return computed property: openQuery', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.openQuery).toBeUndefined();
    wrapper.setData({ openQueryId: '1' });
    expect(wrapper.vm.openQuery).toEqual(testPropsData.queries[0]);
    wrapper.setData({ openQueryId: '3' });
    expect(wrapper.vm.openQuery).toEqual(testPropsData.queries[2]);
  });

  it('should return computed property: closedQuery', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.closedQuery).toBeUndefined();
    wrapper.setData({ closedQueryId: '1' });
    expect(wrapper.vm.closedQuery).toEqual(testPropsData.queries[0]);
    wrapper.setData({ closedQueryId: '3' });
    expect(wrapper.vm.closedQuery).toEqual(testPropsData.queries[2]);
  });

  it('should watch v-model', async () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({
      openQueryId: '1',
      closedQueryId: '2',
      tabIndex: 1,
    });
    await wrapper.vm.$nextTick();
    wrapper.setData({
      value: '1',
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.openQueryId).toEqual(null);
    expect(wrapper.vm.closedQueryId).toEqual(null);
    expect(wrapper.vm.tabIndex).toBe(0);
  });

  it('should have method which set open query id to view: viewOpenQuery', () => {
    const wrapper = createMountedWrapper();
    wrapper.vm.viewOpenQuery('openQueryId');
    expect(wrapper.vm.openQueryId).toBe('openQueryId');
  });

  it('should have method which reset open query id to hide view: hideOpenQuery', () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({
      openQueryId: 'openQueryId',
    });
    wrapper.vm.hideOpenQuery();
    expect(wrapper.vm.openQueryId).toEqual(null);
  });

  it('should have method which set closed query id to view: viewClosedQuery', () => {
    const wrapper = createMountedWrapper();
    wrapper.vm.viewClosedQuery('openQueryId');
    expect(wrapper.vm.closedQueryId).toBe('openQueryId');
  });

  it('should have method which reset closed query id to hide view: hideClosedQuery', () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({
      closedQueryId: 'openQueryId',
    });
    wrapper.vm.hideClosedQuery();
    expect(wrapper.vm.closedQueryId).toEqual(null);
  });

  it('should have method which emit events ("createQuery", "input") on query create: createQuery', () => {
    const wrapper = createMountedWrapper();
    wrapper.vm.createQuery({ comment: 'comment' });
    // assert event has been emitted
    expect(wrapper.emitted().createQuery).toBeTruthy();
    expect(wrapper.emitted().input).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().createQuery.length).toBe(1);
    expect(wrapper.emitted().input.length).toBe(1);
    // assert event payload
    expect(wrapper.emitted().createQuery[0][0]).toEqual({
      fieldId: testPropsData.value,
      formId: testPropsData.formId,
      comment: 'comment',
    });
    expect(wrapper.emitted().input[0][0]).toEqual(null);
  });

  it('should have method which emit "createComment" event on query comment create: onQueryCommentCreate', () => {
    const wrapper = createMountedWrapper();
    wrapper.vm.onQueryCommentCreate('comment');
    // assert event has been emitted
    expect(wrapper.emitted().createComment).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().createComment.length).toBe(1);
    // assert event payload
    expect(wrapper.emitted().createComment[0][0]).toEqual('comment');
  });

  it('should have method which emit "input" event on query cancel: cancelQuery', () => {
    const wrapper = createMountedWrapper();
    wrapper.vm.cancelQuery();
    // assert event has been emitted
    expect(wrapper.emitted().input).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().input.length).toBe(1);
    // assert event payload
    expect(wrapper.emitted().input[0][0]).toEqual(null);
  });

  it('should have method which emit "resetFilter" event on filter reset: resetFilter', () => {
    const wrapper = createMountedWrapper();
    wrapper.vm.resetFilter();
    // assert event has been emitted
    expect(wrapper.emitted().resetFilter).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().resetFilter.length).toBe(1);
  });

  it('should have method which emit "onQueryClosed" event on query close: onQueryClosed', () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({
      openQueryId: 'openQueryId',
    });
    wrapper.vm.onQueryClosed('queryPayload');
    // assert change state
    expect(wrapper.vm.openQueryId).toEqual(null);
    // assert event has been emitted
    expect(wrapper.emitted().onQueryClosed).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().onQueryClosed.length).toBe(1);
    // assert event payload
    expect(wrapper.emitted().onQueryClosed[0][0]).toBe('queryPayload');
  });

  it('should have method which emit "onQueryReopened" event on query reopen: onQueryReopened', () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({
      closedQueryId: 'closedQueryId',
    });
    wrapper.vm.onQueryReopened('queryPayload');
    // assert change state
    expect(wrapper.vm.closedQueryId).toEqual(null);
    // assert event has been emitted
    expect(wrapper.emitted().onQueryReopened).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().onQueryReopened.length).toBe(1);
    // assert event payload
    expect(wrapper.emitted().onQueryReopened[0][0]).toBe('queryPayload');
  });

  it('has the expected html structure', async () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.$el).toMatchSnapshot();
    wrapper.setData({
      openQueryId: '1',
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot();
    wrapper.vm.hideOpenQuery();
    wrapper.setData({
      closedQueryId: '1',
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  it('should switch between tabs', async () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({
      tabIndex: 1,
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot();
  });
});
