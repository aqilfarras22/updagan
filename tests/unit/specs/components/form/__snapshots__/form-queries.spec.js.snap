// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/form/form-queries.vue has the expected html structure 1`] = `
<b-tabs-stub
  class="form-queries-tabs"
  fakeprop=""
  tag="div"
  value="0"
>
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Open Queries: 2"
  >
    <b>
      FormQueriesCommentForm
    </b>
  </b-tab-stub>
   
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Closed Queries: 1"
  >
    <div>
      <!---->
       
      <b
        queries="[object Object]"
      >
        FormQueriesList
      </b>
    </div>
  </b-tab-stub>
</b-tabs-stub>
`;

exports[`components/form/form-queries.vue has the expected html structure 2`] = `
<b-tabs-stub
  class="form-queries-tabs"
  fakeprop=""
  tag="div"
  value="0"
>
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Open Queries: 2"
  >
    <b>
      FormQueriesCommentForm
    </b>
  </b-tab-stub>
   
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Closed Queries: 1"
  >
    <div>
      <!---->
       
      <b
        queries="[object Object]"
      >
        FormQueriesList
      </b>
    </div>
  </b-tab-stub>
</b-tabs-stub>
`;

exports[`components/form/form-queries.vue has the expected html structure 3`] = `
<b-tabs-stub
  class="form-queries-tabs"
  fakeprop=""
  tag="div"
  value="0"
>
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Open Queries: 2"
  >
    <b>
      FormQueriesCommentForm
    </b>
  </b-tab-stub>
   
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Closed Queries: 1"
  >
    <div>
      <b
        closed=""
        query="[object Object]"
      >
        FormQueriesQueryDetails
      </b>
    </div>
  </b-tab-stub>
</b-tabs-stub>
`;

exports[`components/form/form-queries.vue should switch between tabs 1`] = `
<b-tabs-stub
  class="form-queries-tabs"
  fakeprop=""
  tag="div"
  value="1"
>
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Open Queries: 2"
  >
    <b>
      FormQueriesCommentForm
    </b>
  </b-tab-stub>
   
  <b-tab-stub
    buttonid=""
    href="#"
    tag="div"
    title="Closed Queries: 1"
  >
    <div>
      <!---->
       
      <b
        queries="[object Object]"
      >
        FormQueriesList
      </b>
    </div>
  </b-tab-stub>
</b-tabs-stub>
`;
